Firmware/mmu2.cpp

```diff
@@ -299,52 +302,91 @@ void MMU2::ResetRetryAttempts(){
     retryAttempts = MAX_RETRIES;
 }
 
-void MMU2::DecrementRetryAttempts(){
-    if (inAutoRetry && retryAttempts)
-    {
+void MMU2::DecrementRetryAttempts() {
+    if (inAutoRetry && retryAttempts) {
         SERIAL_ECHOLNPGM("DecrementRetryAttempts");
         retryAttempts--;
     }
 }
 
-void MMU2::update_tool_change_counter_eeprom()
+bool MMU2::VerifyFilamentEnteredPTFE()
 {
-    uint32_t toolchanges = eeprom_read_dword((uint32_t*)EEPROM_TOTAL_TOOLCHANGE_COUNT);
-    eeprom_update_dword((uint32_t *)EEPROM_TOTAL_TOOLCHANGE_COUNT, toolchanges + (uint32_t)read_toolchange_counter());
-    reset_toolchange_counter();
+    st_synchronize();
+
+    if (!fsensor.getFilamentPresent()) return false;
+
+    uint8_t fsensorState = 0;
+    // MMU has finished its load, push the filament further by some defined constant length
+    // If the filament sensor reads 0 at any moment, then report FAILURE
+    current_position[E_AXIS] += MMU2_EXTRUDER_PTFE_LENGTH + MMU2_EXTRUDER_HEATBREAK_LENGTH - logic.ExtraLoadDistance();
+    plan_buffer_line_curposXYZE(MMU2_LOAD_TO_NOZZLE_FEED_RATE);
+
+    while(blocks_queued())
+    {
+        // Wait for move to finish and monitor the fsensor the entire time
+        fsensorState |= !fsensor.getFilamentPresent();
+    }
+
+    if (fsensorState)
+    {
+        IncrementLoadFails();
+        return false;
+    } else {
+        // else, happy printing! :)
+        // Revert the movements
+        current_position[E_AXIS] -= (MMU2_EXTRUDER_PTFE_LENGTH + MMU2_EXTRUDER_HEATBREAK_LENGTH - logic.ExtraLoadDistance());
+        plan_buffer_line_curposXYZE(MMU2_LOAD_TO_NOZZLE_FEED_RATE);
+        st_synchronize();
+        return true;
+    }
```

Comment:
You need to check the fsensor during and *after* the unload for this check to be effective. Checking it once after the push isn't really going to help anything.

Reply:
Done: [6ccd375](https://github.com/prusa3d/Prusa-Firmware/pull/3761/commits/6ccd375f006f31137813de471eadb267edd1290e) the filament sensor is checked continously. I am using the binary operator `|=` to set the variable if 1 is seen. I'm assuming in this test that the Fsensor reading is debounced enough elsewhere in our filament sensor code so should reduce risk of false positives.